accepted/system-range/system-range.md

```diff
@@ -0,0 +1,200 @@
+let's say you want to split this data into two tensors, one representing the
+inputs and one the outputs.
+
+This might look like this:
+
+```csharp
+Tensor<float> testDataTable = GetTestData();
+int columnCount = testDataTable.GetLength(1);
+int outputColumn = columnCount - 1;
+
+Tensor<float> inputTable = testDataTable[.., ..outputColumn];
+Tensor<float> outputTable = testDataTable[.., outputColumn..];
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting negative ranges or element types
+  other than `Int32`).
+
+### Non-Goals
+
+* Supporting other numeric types or even offer a generic range type
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+
+## Design
+
+### Proposed API
+
+```csharp
+namespace System {
+    public struct Range : IEnumerable, IEnumerable<int>, IEquatable<Range> {
+        public static Range Empty { get; }
+        public static Range All { get; }
+        public static Range Between(int start, int end);
+        public static Range From(int start);
+        public static Range To(int end);
+        public Range(int index, int length);
+        public int Index { get; }
+        public int Length { get; }
+        public bool IsClosed { get; }
+        public Range Close(Range range);
+        public override string ToString();
+        public override int GetHashCode();
+        public bool Equals(Range other);
+        public override bool Equals(object other);
+        public static bool operator==(Range left, Range right);
+        public static bool operator!=(Range left, Range right);
+    }
+}
+```
+
+### Companion APIs
+
+```csharp
+namespace System {
+    public partial class String {
+        public string this[Range range] { get; }
+    }
+    public partial class T[] {
+        public T[] this[Range range] { get; }
+    }
+    public partial struct ArraySegment<T> {
+        public ArraySegment<T> this[Range range] { get; }
+    }      
+    public partial struct Span<T> {
+        public Span<T> this[Range range] { get; }
+    }
+}
+```
+
+## Q & A
+
+### Open Design Points
+
+* Right now, we cannot retreive `Index` or `Length` if the range is open on any
+  side (it throws `NotSupportedException`).
+    - We could expose `Start: int?` and `End: int?` and then add a factory
+      method `Between(int? start, int? end)`
+
+* We should check whether we can handle negative numbers for `Length` so that we
+  can support addressing relative to the end. Check Python to see what they did
+  there.
+
+* Should `Range` support containment checks?
+    - Positions and/or ranges?
+    - Question is how to define semantics for open ended ranges.
```

Comment:
If we understand open ended ranges as not having an end on one side, then `Range.From(x).Contains(y)` for all `y >= x`, and `Range.To(x).Contains(y)` for all `y < x`. Similarly for ranges:
* `Range.From(x).Contains(Range.Between(y, z))` and `Range.From(x).Contains(Range.From(y))` iff `y >= x`
* `Range.To(x).Contains(Range.Between(y, z))` and `Range.From(x).Contains(Range.To(z))` iff `z < x`
* `Range.To(x).Contains(Range.From(y))` and `Range.From(x).Contains(Range.To(y))` would then never be true, since they might be overlapping but cannot be contained since they both go to infinity on different sides (at least assuming that a `Range.To()` does not implicitly stop at 0).

Reply:
I think the heart of the question is whether we want to define it for partially closed ranges or whether we consistently fail with an exception. That latter might be easier to understand and debug. My knee-jerk reaction is that you shouldn't have partially closed ranges when you perform range checks.